src/capwap_loc_provider.erl

```diff
@@ -0,0 +1,152 @@
+
+%% callbacks
+
+-callback config_fun(any()) -> fun((<<>>) -> {error, any()} | {location, <<>>, <<>>}).
+%% A separate callback returning an opaque ref would be
+%% useful in the case of stateful providers.
+%% This option is not covered yet.
+
+%% user api
+
+start_link(Config) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Config, []).
+
+start(Config) ->
+    gen_server:start({local, ?MODULE}, ?MODULE, Config, []).
+
+%% validation
+
+validate(#{refresh := RefreshTime, providers := Providers})
+  when is_integer(RefreshTime) andalso RefreshTime>0 ->
+    validate_providers(Providers),
+    ok.
+
+validate_providers(Providers) when is_list(Providers) ->
+    lists:map(fun validate_provider/1, Providers),
+    ok.
+
+validate_provider({Mod, Config}) ->
+    Mod:validate_provider(Config).
+
+%% gen_server impl
+
+init(Config = #{refresh := RefreshTime, providers := Providers}) ->
+    validate(Config),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    TabOpts = [set, protected, named_table, {read_concurrency, true}],
+    %% Not needed since it will return the table name
+    LocTab = ets:new(?CACHE_LOC, TabOpts),
+    ?LOG(debug, "cache started: ~p", [LocTab]),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    %% TODO Handle exceptions in chain?
+    {ok, #loc_cache{
+	loc_tab = LocTab, refresh = RefreshTime, providers = Providers, provider_chain = chain(Providers)}}.
+
+handle_call(flush_loc_cache, _, State = #loc_cache{loc_tab = LocTab}) ->
+    ?LOG(debug, "Flush location cache, tableref: ~p", [ets:whereis(LocTab)]),
+    ets:delete_all_objects(LocTab),
+    {reply,ok,State};
+handle_call({get_loc, UseCache = true, Name}, _, State = #loc_cache{
+    loc_tab = LocTab, refresh = RefreshTime, provider_chain = Chain}) ->
+	?LOG(debug, "get location for device name (with cache): ~p", [Name]),
+    Result = case ets:lookup(LocTab, Name) of
+	[{Name, Loc = {location, _, _}}] -> Loc;
+	[] -> refresh_loc(Name, Chain, LocTab, RefreshTime)
+    end,
+    {reply, Result, State};
+handle_call({get_loc, UseCache = false, Name}, _, State = #loc_cache{
+	loc_tab = LocTab, refresh = RefreshTime, provider_chain = Chain}) ->
+    ?LOG(debug, "get location for device name (without cache): ~p", [Name]),
+    Result = Chain(Name),
+    {reply, Result, State};
+handle_call({load_config, #{refresh := RefreshTime, providers := Providers}}, _, OldState) ->
+    NewState = OldState#loc_cache{refresh = RefreshTime, providers = Providers, provider_chain = chain(Providers)},
+    ?LOG(debug, "reloading config: from ~p to ~p", [OldState#loc_cache{provider_chain = undefined}, NewState]),
+    {reply,OldState,NewState}.
+
+handle_cast(_, State) ->
+    ?LOG(error, "Unexpected cast"),
+    {noreply,State}.
+
+handle_info({evict, Name}, State = #loc_cache{loc_tab = LocTab}) ->
+    ?LOG(debug, "refresh location expired for device name: ~p", [Name]),
+    ets:delete(LocTab, Name),
+    {noreply,State}.
+
+terminate(Reason, _) ->
```

Comment:
It maybe make sense or delete ETS here or transfer to other process - in the worst case, after terminate, you cannot create the table with the same name

Reply:
At this stage it is not necessary. If the server dies, the table will die with it. Since it's just a cache, losing its state every now and then due to failures would not be an issue.